### PR TITLE
Inverse Dirichlet Adaptive Loss

### PR DESCRIPTION
**Pull request by hpieper14**
Adds inverse dirichlet adaptive loss function to address [#500](https://github.com/SciML/NeuralPDE.jl/issues/500)

### CODE REVIEW

File: src/pinns_pde_solve.jl [hunk 1]
```diff
@@ -298,6 +298,39 @@ SciMLBase.@add_kwonly function GradientScaleAdaptiveLoss(reweight_every; weight_
...
+
+from paper
+Inverse Dirichlet weighting enables reliable training of physics informed neural networks
+Suryanarayana Maddu, Dominik Sturm, Christian L Müller, and Ivo F Sbalzarini
+https://iopscience.iop.org/article/10.1088/2632-2153/ac3712/pdf
+with code reference
+https://github.com/mosaic-group/inverse-dirichlet-pinn
+"""
+
+mutable struct InverseDirichletAdaptiveLoss{T <: Real} <: AbstractAdaptiveLoss
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:319]:
```suggestion
"""
mutable struct InverseDirichletAdaptiveLoss{T <: Real} <: AbstractAdaptiveLoss
```

No space between the docstring and the structs. Otherwise it won't show up in the help.

File: src/pinns_pde_solve.jl [hunk 2]
```diff
@@ -298,6 +298,39 @@ SciMLBase.@add_kwonly function GradientScaleAdaptiveLoss(reweight_every; weight_
...
+    reweight_every::Int64
+    weight_change_inertia::T
+    pde_loss_weights::Vector{T} 
+    bc_loss_weights::Vector{T} 
+    additional_loss_weights::Vector{T} 
+    SciMLBase.@add_kwonly function InverseDirichletAdaptiveLoss{T}(reweight_every; weight_change_inertia=0.9, pde_loss_weights=1, bc_loss_weights=1, additional_loss_weights=1) where T <: Real
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:324]: ` weight_change_inertia=0.9`, what's the justification on that?

File: src/pinns_pde_solve.jl [hunk 3]
```diff
@@ -1406,6 +1439,25 @@ function discretize_inner_functions(pde_system::PDESystem, discretization::Physi
             end
             nothing
         end
+    if adaloss isa InverseDirichletAdaptiveLoss
+        weight_change_inertia = discretization.adaptive_loss.weight_change_inertia
+        function run_loss_inverse_dirichlet_adaptive_loss(0)
+            if iteration[1] % adaloss.reweight_every == 0
+                pde_grads_std_all = [std(Zygote.gradient(pde_loss_function, 0)[1]) for pde_loss_function in pde_loss_function]
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:1445]:
```suggestion
                pde_grads_std_all = [std(Zygote.gradient(pde_loss_function, 0)[1]) for pde_loss_function in pde_loss_functions]
```

File: src/pinns_pde_solve.jl [hunk 4]
```diff
@@ -1406,6 +1439,25 @@ function discretize_inner_functions(pde_system::PDESystem, discretization::Physi
...
+                pde_grads_std_max = maximum(pde_grads_std_all)
+                bc_grads_std = [std(Zygote.gradient(bc_loss_function, 0)[1]) for bc_loss_function in bc_loss_funcitons]
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:1447]:
```suggestion
                bc_grads_std = [std(Zygote.gradient(bc_loss_function, 0)[1]) for bc_loss_function in bc_loss_functions]
```

File: src/pinns_pde_solve.jl [hunk 5]
```diff
@@ -298,6 +298,39 @@ SciMLBase.@add_kwonly function GradientScaleAdaptiveLoss(reweight_every; weight_
...
+from paper
+Inverse Dirichlet weighting enables reliable training of physics informed neural networks
+Suryanarayana Maddu, Dominik Sturm, Christian L Müller, and Ivo F Sbalzarini
+https://iopscience.iop.org/article/10.1088/2632-2153/ac3712/pdf
+with code reference
+https://github.com/mosaic-group/inverse-dirichlet-pinn
+"""
+
+mutable struct InverseDirichletAdaptiveLoss{T <: Real} <: AbstractAdaptiveLoss
+    reweight_every::Int64
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:319]: Rename this `reweights_steps`? That's a bit more consistent with the rest of the ecosystem

File: src/pinns_pde_solve.jl [hunk 6]
```diff
@@ -1406,6 +1439,25 @@ function discretize_inner_functions(pde_system::PDESystem, discretization::Physi
...
         end
+    if adaloss isa InverseDirichletAdaptiveLoss
+        weight_change_inertia = discretization.adaptive_loss.weight_change_inertia
+        function run_loss_inverse_dirichlet_adaptive_loss(0)
+            if iteration[1] % adaloss.reweight_every == 0
+                pde_grads_std_all = [std(Zygote.gradient(pde_loss_function, 0)[1]) for pde_loss_function in pde_loss_function]
+                pde_grads_std_max = maximum(pde_grads_std_all)
+                bc_grads_std = [std(Zygote.gradient(bc_loss_function, 0)[1]) for bc_loss_function in bc_loss_funcitons]
+
+                nonzero_divisor_eps =  adaloss_T isa Float64 ? Float64(1e-11) : convert(adaloss_T, 1e-7)
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:1449]: What's the justification on the `1e-11` and `1e-7`?

**Comment by ChrisRackauckas** [comment]
The biggest thing is that this needs tests.

File: src/pinns_pde_solve.jl (hunk 2 above)
Comment (hpieper14) [reply src/pinns_pde_solve.jl:324]: I wasn't sure what to choose as default values for constants (see also `nonzero_divisor_eps`) so I choose the same ones as in `GradientScaleAdaptiveLoss`. What's the best way to choose these?

Comment (ChrisRackauckas) [reply src/pinns_pde_solve.jl:324]: Is there a default from the original code? If not, we should keep it required unless we find a value that tends to work in 90% of cases.

Comment (hpieper14) [reply src/pinns_pde_solve.jl]: Changed in commit [cec87ef](https://github.com/SciML/NeuralPDE.jl/pull/504/commits/cec87ef6828f5a410c9e652ea309d6800d7a1de9).

**Comment by ChrisRackauckas** [comment]
Try the incompressible Navier-Stokes from the paper. I think this may only make a difference when it's "sufficiently hard", and Poisson is simple enough that non-adaptive is fine.

File: src/pinns_pde_solve.jl (hunk 6 above)
Comment (ChrisRackauckas) [reply src/pinns_pde_solve.jl:1452]: Bump on this.

Comment (hpieper14) [reply src/pinns_pde_solve.jl:1452]: I looked over the implementation by the authors of the paper and they don't include an additional epsilon in the denominator, so I think I should remove this.

File: src/pinns_pde_solve.jl (hunk 5 above)
Comment (hpieper14) [reply src/pinns_pde_solve.jl:319]: I chose `reweight_every` to be consistent with the variable names chosen for `MiniMaxAdaptiveLoss` and `GradientScaleAdaptiveLoss`. Is the naming convention is `reweight_steps` across other files?

**Comment by hpieper14** [comment]
I've started working on a tutorial/test for the incompressible Navier-Stokes and I'll add the tutorial to [docs/src/pinn](https://github.com/SciML/NeuralPDE.jl/tree/master/docs/src/pinn). Should I add the test to [adaptive_loss_tests.jl](https://github.com/SciML/NeuralPDE.jl/blob/master/test/adaptive_loss_tests.jl) too?

**Comment by ChrisRackauckas** [comment]
>Should I add the test to [adaptive_loss_tests.jl](https://github.com/SciML/NeuralPDE.jl/blob/master/test/adaptive_loss_tests.jl) too?

If it's not too long. That might get constrained by compute time when used in the test infrastructure.

File: src/pinns_pde_solve.jl (hunk 5 above)
Comment (ChrisRackauckas) [reply src/pinns_pde_solve.jl:319]: `progress_steps`, `timeseries_steps`, `p_steps`, etc. It's not the biggest deal, but it's slightly more consistent with what else is out there in SciML.

File: src/pinns_pde_solve.jl [hunk 7]
```diff
@@ -298,6 +298,38 @@ SciMLBase.@add_kwonly function GradientScaleAdaptiveLoss(reweight_every; weight_
...
+    additional_loss_weights::Vector{T} 
+    SciMLBase.@add_kwonly function InverseDirichletAdaptiveLoss{T}(reweight_every; weight_change_inertia=0.5, pde_loss_weights=1, bc_loss_weights=1, additional_loss_weights=1) where T <: Real
+        new(convert(Int64, reweight_every), convert(T, weight_change_inertia), vectorify(pde_loss_weights, T), vectorify(bc_loss_weights, T), vectorify(additional_loss_weights, T))
+    end
+end
+# default to Float64
+SciMLBase.@add_kwonly function InverseDirichletAdaptiveLoss(reweight_every; weight_change_inertia=0.5, pde_loss_weights=1, bc_loss_weights=1, additional_loss_weights=1) 
+    InverseDirichletAdaptiveLoss{Float64}(reweight_every; weight_change_inertia=weight_change_inertia, 
+        pde_loss_weights=pde_loss_weights, bc_loss_weights=bc_loss_weights, additional_loss_weights=additional_loss_weights)
+end
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:332]:
```suggestion
InverseDirichletAdaptiveLoss(args...; kwargs...) = InverseDirichletAdaptiveLoss{Float64}(args...; kwargs...)
```

**Review by ChrisRackauckas — APPROVED** [review]
This looks right to be. @zoemcc ?

**Comment by zoemcc** [comment]
This looks really good but I don't have time until my project deadline on Tuesday to review further.

**Comment by ChrisRackauckas** [comment]
This should get rebased due to changes in https://github.com/SciML/NeuralPDE.jl/pull/553. That would hopefully make it much cleaner too.
